### DISCUSSION TITLE
docker-acap  →　 acap-computer-vision-sdk-examples　（object-detector-python）　where is the problem?

### DISCUSSION THREAD

**Discussion by hfgsdjkfhkjsdfj** [AxisCommunications/acap-computer-vision-sdk-examples#133] [Q&A]
NAME="Ubuntu"
VERSION_ID="22.04"

Docker version 20.10.22, build 3a2c30b

#互換性の確認
DEVICE_IP=192.168.10.159
DEVICE_PASSWORD='test33'

curl -s --anyauth -u "root:$DEVICE_PASSWORD" \
  "http://$DEVICE_IP/axis-cgi/param.cgi?action=update&root.Network.SSH.Enabled=yes"

ssh root@$DEVICE_IP 'command -v containerd >/dev/null 2>&1 && echo Compatible with Docker ACAP || echo Not compatible with Docker ACAP'

# インストール
docker run --rm axisecp/docker-acap:1.3.0-armv7hf 192.168.10.1 install

#TLSを使用したドッカーACAPの固定
DEVICE_IP=192.168.10.159
DEVICE_PASSWOR'

curl -s --anyauth -u "root:$DEVICE_PASSWORD" \
  "http://$DEVICE_IP/axis-cgi/param.cgi?action=update&root.dockerdwrapper.UseTLS=yes"

#Dockerデーモンソケットを保護するためにTLS ( HTTPS )を使用します
#まず、CA 秘密鍵と公開鍵を作成します。

openssl genrsa -aes256 -out ca-key.pem 4096
#ke

openssl req -new -x509 -days 365 -key ca-key.pem -sha256 -out ca.pem
#ke HOSTは　MyComputer

openssl genrsa -out server-key.pem 4096

openssl req -subj "/CN=MyComputer" -sha256 -new -key server-key.pem -out server.csr

#公開鍵を CA で署名しましょう
echo subjectAltName = DNS:MyComputer,IP:192.168.10.110,IP:127.0.0.1 >> extfile.cnf

echo extendedKeyUsage = serverAuth >> extfile.cnf

openssl x509 -req -days 365 -sha256 -in server.csr -CA ca.pem -CAkey ca-key.pem \
  -CAcreateserial -out server-cert.pem -extfile extfile.cnf

#クライアント認証用に、クライアント鍵と証明書署名要求を作成します
openssl genrsa -out key.pem 4096

openssl req -subj '/CN=client' -new -key key.pem -out client.csr

#追加　
echo extendedKeyUsage = clientAuth >> extfile.cnf
openssl x509 -req -days 365 -sha256 -in client.csr -CA ca.pem -CAkey ca-key.pem \
  -CAcreateserial -out cert.pem -extfile extfile-client.cnf
#クライアント認証用の鍵を実装するには、追加設定ファイルを作成します。
echo extendedKeyUsage = clientAuth > extfile.cnf

#次は公開鍵に署名します。
openssl x509 -req -days 365 -sha256 -in client.csr -CA ca.pem -CAkey ca-key.pem \
  -CAcreateserial -out cert.pem -extfile extfile-client.cnf

#TLSセットアップ
scp ca.pem server-cert.pem server-key.pem root@192.168.10.159:/usr/local/packages/dockerdwrapper/

#クライアントキーと証明書
DOCKER_PORT=2376
docker --tlsverify \
       --tlscacert=ca.pem \
       --tlscert=cert.pem \
       --tlskey=key.pem \
       -H=192.168.10.159:$DOCKER_PORT \
       version

#各Dockerコマンドでファイルを指定するのはすぐに面倒になります。Dockerを構成するには キーと証明書を自動的に使用します DOCKER_CERT_PATH 環境変数:
export DOCKER_CERT_PATH=/usr/local/packages/dockerdwrapper/
DOCKER_PORT=2376
docker --tlsverify \
       -H=192.168.10.159:$DOCKER_PORT \
       version

#Docker ACAPの使#Docker ACAPの使用
DOCKER_INSECURE_PORT=2376
docker -H=192.168.10.159:$DOCKER_INSECURE_PORT version

#ドッカーACAPの構築
./build.sh armv7hf

docker run --rm docker-acap:1.0 192.168.10.1 install

###Docker ACAP おわり

###acap-computer-vision-sdk-examples/object-detector-python

#アーキテクチャの環境変数をエクスポートする
export ARCH=armv7hf
export CHIP=tpu

# Define app name
export APP_NAME=acap4-object-detector-python
export MODEL_NAME=acap-dl-models

# Install qemu to allow build flask for a different architecture
docker run --rm --privileged multiarch/qemu-user-static --credential yes --persistent yes

# Build app
docker build --tag $APP_NAME --build-arg ARCH .

# Build inference model
docker build --file Dockerfile.model --tag $MODEL_NAME --build-arg ARCH .

#デバイスのIPアドレスを設定し、Dockerメモリをクリアします
DEVICE_IP=192.168.10.159
DOCKER_PORT=2376

docker --tlsverify --host tcp://$DEVICE_IP:$DOCKER_PORT system prune --all --force

#画像をインストールする
docker save $APP_NAME | docker --tlsverify --host tcp://$DEVICE_IP:$DOCKER_PORT load

docker save $MODEL_NAME | docker --tlsverify --host tcp://$DEVICE_IP:$DOCKER_PORT load

#コンテナを実行します
docker-compose --tlsverify --host tcp://$DEVICE_IP:$DOCKER_PORT --env-file ./config/env.$ARCH.$CHIP up

# Terminate with Ctrl-C and cleanup
docker-compose --tlsverify --host tcp://$DEVICE_IP:$DOCKER_PORT --env-file ./config/env.$ARCH.$CHIP down --volumes

**Comment by Corallo** [comment]
Hello @hfgsdjkfhkjsdfj 

At which point of these steps you are having issues?
And what is the error that you get?

**Reply by hfgsdjkfhkjsdfj** [reply]
`camera is AXIS Q1615 Mk III 11.2.68

..................................
root@MyComputer:/docker-acap# scp ca.pem server-cert.pem server-key.pem root@192.168.10.159:/usr/local/packages/dockerdwrapper/
root@192.168.10.159's password:
ca.pem 100% 2049 1.5MB/s 00:00
server-cert.pem 100% 1980 1.3MB/s 00:00
server-key.pem 100% 3272 2.6MB/s 00:00
root@MyComputer:/docker-acap# DOCKER_PORT=2376
root@MyComputer:~/docker-acap# docker --tlsverify
--tlscacert=ca.pem
--tlscert=cert.pem
--tlskey=key.pem
-H=192.168.10.159:$DOCKER_PORT
version
Cannot connect to the Docker daemon at tcp://192.168.10.159:2376. Is the docker daemon running?
↑　issues

Client: Docker Engine - Community
Cloud integration: v1.0.29
Version: 20.10.22
API version: 1.41
Go version: go1.18.9
Git commit: 3a2c30b
Built: Thu Dec 15 22:28:22 2022
OS/Arch: linux/amd64
Context: default
Experimental: true
root@MyComputer:/docker-acap# export DOCKER_CERT_PATH=/usr/local/packages/dockerdwrapper/
DOCKER_PORT=2376
root@MyComputer:/docker-acap# docker --tlsverify
-H=192.168.10.159:$DOCKER_PORT
version
unable to resolve docker endpoint: open /usr/local/packages/dockerdwrapper/ca.pem: no such file or directory
↑issues
root@MyComputer:/docker-acap# cd
root@MyComputer:# docker --tlsverify
-H=192.168.10.159:$DOCKER_PORT
version
unable to resolve docker endpoint: open /usr/local/packages/dockerdwrapper/ca.pem: no such file or directory
↑issues
root@MyComputer:~# DOCKER_INSECURE_PORT=2376
docker -H=192.168.10.159:$DOCKER_INSECURE_PORT version
Cannot connect to the Docker daemon at tcp://192.168.10.159:2376. Is the docker daemon running?`

**Reply by Corallo** [reply]
Right, first of all verify that the docker acap is running on the device.
You can go in the camera page (192.168.10.159 on your browser) and then in app find the docker acap, and see the log.
Check what it says there.
Either the acap is was not started, or some error stopped it. It might also be some connection issue, like the 2376 port being not open.

**Comment by hfgsdjkfhkjsdfj** [comment]
`https://docs.docker.com/engine/security/protect-access/

.......................
root@MyComputer:/docker-acap# openssl genrsa -out key.pem 4096
root@MyComputer:/docker-acap# openssl req -subj '/CN=client' -new -key key.pem -out client.csr
root@MyComputer:/docker-acap# echo extendedKeyUsage = clientAuth > extfile.cnf
root@MyComputer:/docker-acap# openssl x509 -req -days 365 -sha256 -in client.csr -CA ca.pem -CAkey ca-key.pem
-CAcreateserial -out cert.pem -extfile extfile-client.cnf
Can't open "extfile-client.cnf" for reading, No such file or directory
40E7BD9B4C7F0000:error:80000002:system library:BIO_new_file:No such file or directory:../crypto/bio/bss_file.c:67:calling fopen(extfile-client.cnf, r)
40E7BD9B4C7F0000:error:10000080:BIO routines:BIO_new_file:no such file:../crypto/bio/bss_file.c:75:

`

Why issue ???

**Comment by hfgsdjkfhkjsdfj** [comment]
Securing the Docker ACAP using TLS
The Docker ACAP can be run either unsecured or in TLS mode. The Docker ACAP uses TLS as default. Use the "Use TLS" dropdown in the web interface to switch between the two different modes. It's also possible to toggle this option by calling the parameter management API in [VAPIX](https://www.axis.com/vapix-library/) and setting the root.dockerdwrapper.UseTLS parameter to yes or no. The following commands would enable TLS:


`DEVICE_IP=<device` ip>
DEVICE_PASSWORD='<password>'

curl -s --anyauth -u "root:$DEVICE_PASSWORD" \
  "http://$DEVICE_IP/axis-cgi/param.cgi?action=update&root.dockerdwrapper.U

↓

#Use TLS (HTTPS) to protect the Docker daemon socket


`openssl genrsa -aes256 -out ca-key.pem 4096

openssl req -new -x509 -days 365 -key ca-key.pem -sha256 -out ca.pem

openssl genrsa -out server-key.pem 4096

openssl req -subj "/CN=$HOST" -sha256 -new -key server-key.pem -out server.csr

echo subjectAltName = DNS:$HOST,IP:10.10.10.20,IP:127.0.0.1 >> extfile.cnf

echo extendedKeyUsage = serverAuth >> extfile.cnf

openssl x509 -req -days 365 -sha256 -in server.csr -CA ca.pem -CAkey ca-key.pem \
  -CAcreateserial -out server-cert.pem -extfile extfile.cnf

openssl genrsa -out key.pem 4096

openssl req -subj '/CN=client' -new -key key.pem -out client.csr

echo extendedKeyUsage = clientAuth > extfile-client.cnf

openssl x509 -req -days 365 -sha256 -in client.csr -CA ca.pem -CAkey ca-key.pem \
  -CAcreateserial -out cert.pem -extfile extfile-client.cnf

rm -v client.csr server.csr extfile.cnf extfile-client.cnf

chmod -v 0400 ca-key.pem key.pem server-key.pem

chmod -v 0444 ca.pem server-cert.pem cert.pem

↓

scp ca.pem server-cert.pem server-key.pem root@<device ip>:/usr/local/packages/dockerdwrapper/
`

O,K？

**Reply by hfgsdjkfhkjsdfj** [reply]
root@MyComputer:~/docker-acap# scp ca.pem server-cert.pem server-key.pem root@<device ip>:/usr/local/packages/dockerdwrapper/
-bash: device: No such file or directory

ｗｈｙ？

**Reply by Corallo** [reply]
Can you check that the files:
`ca.pem server-cert.pem server-key.pem`
are in your docker-acap folder?
